File: Diagrammatize.Design.Pattern/code/cxx/Visitor/A1/FileFindVisitor.cpp
#include "FileFindVisitor.hpp"
#include "string_algorithm.hpp"

FileFindVisitor::FileFindVisitor(const std::string& filetype_) : filetype(filetype_) {
}

FileFindVisitor::~FileFindVisitor() {
}

void FileFindVisitor::visit(File* file) {
    if (ends_with(file->getName(), filetype)) {
        found.push_back(file);
    }
}

void FileFindVisitor::visit(Directory* directory) {
    for (auto entry : *directory) {
        entry->accept(this);
    }
}


const std::vector<File*>& FileFindVisitor::getFoundFiles() {
    return found;
}

File: Diagrammatize.Design.Pattern/code/python/Visitor/A1/FileFindVisitor.py

from Visitor import Visitor

class FileFindVisitor(Visitor):
    def __init__(self, filetype):
        self.filetype = filetype
        self.found = list()

    def getFoundFiles(self):
        return iter(self.found)

    def visitFile(self, file_):
        if file_.getName().endswith(self.filetype):
            self.found.append(file_)

    def visitDirectory(self, directory):
        for entry in directory:
            entry.accept(self)
